File: src/Python/840.MagicSquaresInGrid.py

class Solution:
    def numMagicSquaresInside(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        count = 0
        for i in range(len(grid) - 2):
            for j in range(len(grid[0]) - 2):
                s = set()
                for m in range(3):
                    for n in range(3):
                        if grid[i + m][j + n] > 9 or grid[i + m][j + n] < 1:
                            break
                        s.add(grid[i + m][j + n])
                if len(s) != 9:
                    continue
                summ = sum(grid[i][j : j + 3])
                if sum(grid[i + 1][j : j + 3]) != summ:
                    continue 
                if sum(grid[i + 2][j : j + 3]) != summ:
                    continue
                if grid[i][j] + grid[i + 1][j] + grid[i + 2][j] != summ:
                    continue
                if grid[i][j + 1] + grid[i + 1][j + 1] + grid[i + 2][j + 1] != summ:
                    continue
                if grid[i][j + 2] + grid[i + 1][j + 2] + grid[i + 2][j + 2] != summ:
                    continue
                if grid[i][j] + grid[i + 1][j + 1] + grid[i + 2][j + 2] != summ:
                    continue
                if grid[i + 2][j] + grid[i + 1][j + 1] + grid[i][j + 2] != summ:
                    continue
                count += 1
        return count


if __name__ == '__main__':
    solution = Solution()
    print(solution.numMagicSquaresInside([[4,3,8,4], [9,5,1,9], [2,7,6,2]]))
    print(solution.numMagicSquaresInside([[10,3,5],[1,6,11],[7,9,2]]))
    print(solution.numMagicSquaresInside([[3,2,9,2,7],[6,1,8,4,2],[7,5,3,2,7],[2,9,4,9,6],[4,3,8,2,5]]))
else:
    pass

File: src/Java/840.MagicSquaresInGrid.java

import java.awt.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;

class Solution {
    public int numMagicSquaresInside(int[][] grid) {
        if (grid.length < 3) { return 0; }
        int num = 0;
        for (int i = 0; i < grid.length-2; ++i) {
            for (int j = 0; j < grid.length-2; ++j) {
                if (isMagicSquares(grid, i, j, 3))
                    ++num;
            }
        }
        return num;
    }

    boolean isMagicSquares(int[][] grid, int i, int j, int n) {
        if (grid.length < i + n || grid.length < j + n) 
            return false;    
        HashSet<Integer> hash = new HashSet<>();
        int val = -1;
        int d1=0,d2=0;
        for (int k = 0; k < 3; ++k) {

            int row = 0, col = 0;
            for (int l = 0; l < 3; ++l) {
                if (grid[i+k][j+l] < 1 || grid[i+k][j+l] > 9)
                    return false;
                hash.add(grid[i+k][j+l]);
                row += grid[i+k][j+l];
                col += grid[i+l][j+k];

                if (k == l) d1 += grid[i+k][j+l];
                if (l+1+k == n) d2 += grid[i+k][j+l];
            }

            if (val< 0) val = row;
            if (row != val || col != val) return false;

        }

        if (val != d1 || val != d2) return false;
        return hash.size() == 9;
    }
}
